spark3/src/main/java/org/apache/iceberg/spark/source/SparkWrite.java

```diff
@@ -262,10 +262,16 @@ public void commit(WriterCommitMessage[] messages) {
   private class DynamicOverwrite extends BaseBatchWrite {
     @Override
     public void commit(WriterCommitMessage[] messages) {
+      Iterable<DataFile> files = files(messages);
+      if (Iterables.size(files) == 0) {
```

Comment:
What about using `!files.hasNext` instead? I'm not sure we want to assume that the iterable can be consumed multiple times.